### ISSUE TITLE
VolumnOp was not able to create PVC

### ISSUE THREAD

**Issue by yuhuishi-convect** [kubeflow/pipelines#5257]
### What steps did you take:

A simple pipeline with one vol_op and one simple step that mounts the pvc created.

```python
# minimal example

import kfp

@kfp.dsl.pipeline(
    name="data download and upload",
    description="Data IO test"
)
def volume_pipeline():
    # shared vol
    
    vop = kfp.dsl.VolumeOp(
        name="volume_creation",
        resource_name="sharedpvc",
        size="5Gi",
        modes=["RWO"]
    )
    
    # mount the vol
    
    simple_task = kfp.dsl.ContainerOp(
        name="simple task",
        image="bash",
        arguments=[
            "echo",
            "hello",
            ">/data/hello.text"
        ]
    ).add_pvolumes({
        "/data": vop.volume
    })
    
# run the pipeline

client = kfp.Client()

client.create_run_from_pipeline_func(volume_pipeline, arguments={})
```

### What happened:
The VolumeOp was not able to create the PVC, therefore the depending task complains about not finding the PVC.
`kubectl get pvc -n kubeflow | grep sharedpvc` didn't return any results. 

### What did you expect to happen:
The VolumeOp shall create a PVC named `sharedpvc`.

### Environment:

How did you deploy Kubeflow Pipelines (KFP)?
Deploying Kubeflow Pipelines on a local kind cluster

KFP version: 1.2.0 

KFP SDK version: 1.4.0 

### Anything else you would like to add:
[Miscellaneous information that will assist in solving the issue.]

The log of the VolumneOp indicates
```
This step output is taken from cache.
```
I was trying to prevent it from using the cache but didn't succeed.
The manifest from the VolumneOp
```
apiVersion: v1 kind: PersistentVolumeClaim metadata: name: '{{workflow.name}}-sharedpvc' spec: accessModes: - RWO resources: requests: storage: 5Gi 
```

The log of the depending task indicates
```
This step is in Pending state with this message: Unschedulable: persistentvolumeclaim "{{tasks.volume-creation.outputs.parameters.volume-creation-name}}" not found
```

Storage class used
``` yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"storage.k8s.io/v1","kind":"StorageClass","metadata":{"annotations":{"storageclass.kubernetes.io/is-default-class":"true"},"name":"standard"},"provisioner":"rancher.io/local-path","reclaimPolicy":"Delete","volumeBindingMode":"WaitForFirstConsumer"}
    storageclass.kubernetes.io/is-default-class: "true"
  creationTimestamp: "2020-12-29T06:31:17Z"
  name: standard
  resourceVersion: "195"
  selfLink: /apis/storage.k8s.io/v1/storageclasses/standard
  uid: 5dbc1bff-b488-4d3a-b45f-e710cf96a415
provisioner: rancher.io/local-path
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer

```

/kind bug

**Comment by Bobgy** [comment]
/assign @elikatsis 
Can you help with this issue?

**Comment by munagekar** [comment]
Kubeflow cache's steps, so given the same inputs, it skips the step and instead fetches outputs from minio. Volume op has not outputs in Minio so ideally VolumeOp should not be cached. In this case you would need to change the name of the PVC. 

You can also refer to https://github.com/kubeflow/pipelines/issues/5055#issue-796655777.

**Comment by munagekar** [comment]
Does the following work ?
```python
    vop = kfp.dsl.VolumeOp(
        name="volume_creation",
        resource_name=f"{{{{workflow.name}}}}-sharedpvc",
        size="5Gi",
        modes=["RWO"]
    )
```

This will hopefully change the input to the ResourceOp and hence prevent caching.

**Comment by yuhuishi-convect** [comment]
> 
> 
> Does the following work ?
> 
> ```python
>     vop = kfp.dsl.VolumeOp(
>         name="volume_creation",
>         resource_name=f"{{{{workflow.name}}}}-sharedpvc",
>         size="5Gi",
>         modes=["RWO"]
>     )
> ```
> 
> This will hopefully change the input to the ResourceOp and hence prevent caching.

Yes, the problem is fixed after I disabled the cache. Thanks for the help.

**Comment by elikatsis** [comment]
@Bobgy sorry I had totally missed this.

I think the problem here is that
1. the mechanism is caching steps that it shouldn't do so
2. there is no "user" selection on whether to cache some specific step or not, only global API server configuration. The API server overrides any configuration: https://github.com/kubeflow/pipelines/blob/601b104c0beba8049dfee493d0b07c0aea390b78/backend/src/apiserver/resource/resource_manager.go#L377

**Comment by elikatsis** [comment]
I'll reopen this issue as it needs some fix apart from globally disabling the cache

/reopen

**Comment by pahask8** [comment]
> Does the following work ?
> 
> ```python
>     vop = kfp.dsl.VolumeOp(
>         name="volume_creation",
>         resource_name=f"{{{{workflow.name}}}}-sharedpvc",
>         size="5Gi",
>         modes=["RWO"]
>     )
> ```
> 
> This will hopefully change the input to the ResourceOp and hence prevent caching.

Hi, for some reason, this workaround does not work for me. KF version: 1.6.0, Kubeflow  SDK 1.6.3. I used the same pipeline from the issue description. My pcv was created only once.

Thank you

**Comment by mjurkus** [comment]
Facing the same issue. `VolumeOp` fails to create PVC with log message `This step output is taken from cache.`

Tried to change `resource_name`, but that workaround didn't work (if it worked before).

**Comment by Bobgy** [comment]
I think KFP v1 caching should not cache volume op / resource op, because the side effect is intended.
/cc @Ark-kun 

Welcome contributions to fix this.
Caching webhook: https://github.com/kubeflow/pipelines/tree/master/backend/src/cache.

Besides that, caching for KFP v2 compatible mode should no longer cache volume ops. You can also consider trying it out too when it's released and documented for your env. (currently documented for KFP standalone, but not full Kubeflow.)

**Comment by boarder7395** [comment]
Not sure this is the correct place to bring this up. And I am not familiar with v2 component configuration yet but it looks like the mutation webhook in cache-server backend is looking for key pipelines.kubeflow.org/enable_caching while the python sdk is creating key pipelines.kubeflow.org/cache_enabled when using set_cache_enabled.

See following files:
https://github.com/kubeflow/pipelines/blob/master/backend/src/cache/server/mutation.go#L36
https://github.com/kubeflow/pipelines/blob/dec03067ca1f89f1ca23c7397830d60201448fa6/sdk/python/kfp/compiler/_op_to_template.py#L186

Although it was mentioned to use .add_pod_annotation(name="pipelines.kubeflow.org/max_cache_staleness", value="P0D") which looks like it should work from the code in mutation.go. If that's the case will the base_op function set_caching_enabled be deprecated in the future. 

The change on cache-server to allow the function of set_caching_enabled to work will be easy. I made the changes and tested them on my forked version of this repo. I can make a PR if the set_caching_enabled is not going to be deprecated.

**Comment by shashisingh** [comment]
What is the solution for this issue in version 1.8.0 of kubeflow while using 1.8.22 version of kfp sdk? What I want is to be able to run multiple instance of a pipeline (these are model training jobs) and want to make sure persistent volumes are not shared between these jobs. I have tried the following so far =>
 ```
volumeOp.add_pod_label("pipelines.kubeflow.org/enable_caching","false") 
        .add_pod_label("pipelines.kubeflow.org/cache_enabled","false")
        .set_caching_options(enable_caching=False)
``` 
I have also tried to disable cache globally by following 
https://www.kubeflow.org/docs/components/pipelines/legacy-v1/overview/caching/#disabling-caching-in-your-kubeflow-pipelines-deployment
but I can still see pvcs are still being reused across different runs.

**Comment by juliusvonkohout** [comment]
@shashisingh the solution is in this thread and confirmed by others

/close
